libkrun X11 pipewire support

dorindabassey opened the pull request:
This series adds cross-domain IPC support for X11 and PipeWire, enabling better graphics with enhanced file descriptor and futex sharing capabilities.

WhatAmISupposedToPutHere committed:
Add X11 cross domain channel type.

This adds the X11 cross domain channel and an ability to share
memory watchers between the VM and the host for synchronization.
Memory watchers provide a platform-agnostic API over Linux
futex syscalls, allowing shared memory synchronization via
VirtioFS file descriptors.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>
Co-developed-by: Asahi Lina <lina@asahilina.net>
Signed-off-by: Asahi Lina <lina@asahilina.net>
Signed-off-by: Dorinda Bassey <dbassey@redhat.com>

hoshinolina committed:
Add virtiofs_table support to cross-domain

This adds the VirtiofsTable type and plumbs it
through RutabagaBuilder and CrossDomain, allowing
file descriptors to be shared between virtio-fs
and virtio-gpu contexts.

Based on libkrun commit 4ccc22d2ff887ae22eeeb9d0434498209dfdc00c

Signed-off-by: Asahi Lina <lina@asahilina.net>
Signed-off-by: Dorinda Bassey <dbassey@redhat.com>

gurchetansingh reviewed (changes requested):
O_o ... this is a tricky one.  I think the core thing you must try to do here is try to fit the cross-platform abstraction we're trying to create.

* You should use `rustix`, not `libc` or `nix`
* For `EVENT_FD`, we already have an abstraction for `mesa3d_util::Event`.  Use that.
* For `futex`, you need to create that.  The good thing is:

https://github.com/m-ou-se/atomic-wait

has a pretty solid API for this over Linux (futex), Windows (WaitOnAddress), and MacOS.  Do the Linux implementation, but we'll need stubs for Windows/MacOS (hopefully eventually we'll get to them too).  Use `rustix` instead of `libc` (most of your changes will be `mesa3d_util`).  We of course need to properly attribute `atomic-wait` for the adaptation, so I think the resultant code needs to be BSD-2 licensed.

And also, can you test this change with `muvm` + `libkrun` + latest rutabaga_gfx?  That'll help us have confidence the change works before cut the release.

Thank you!

Commit:
Add pipewire cross domain channel type.

This adds a pipewire channel type, increases the maximum number
of fds that can be transferred in a single message, and adds support
for sharing eventfds between the guest and host.

Signed-off-by: Sasha Finkelstein <fnkl.kernel@gmail.com>
Signed-off-by: Dorinda Bassey <dbassey@redhat.com>

dorindabassey commented:
> O_o ... this is a tricky one. I think the core thing you must try to do here is try to fit the cross-platform abstraction we're trying to create.
> 
>     * You should use `rustix`, not `libc` or `nix`
> 
>     * For `EVENT_FD`, we already have an abstraction for `mesa3d_util::Event`.  Use that.
> 
>     * For `futex`, you need to create that.  The good thing is:
> 
> 
> https://github.com/m-ou-se/atomic-wait
> 
> has a pretty solid API for this over Linux (futex), Windows (WaitOnAddress), and MacOS. Do the Linux implementation, but we'll need stubs for Windows/MacOS (hopefully eventually we'll get to them too). Use `rustix` instead of `libc` (most of your changes will be `mesa3d_util`). We of course need to properly attribute `atomic-wait` for the adaptation, so I think the resultant code needs to be BSD-2 licensed.
> 
> And also, can you test this change with `muvm` + `libkrun` + latest rutabaga_gfx? That'll help us have confidence the change works before cut the release.
> 
> Thank you!

Thank you for the feedback, I have addressed the issues and incorporated your suggestions to use rustix. I also tested these changes with `muvm` from (git clone -b val/kqloytrpuwkm https://github.com/valpackett/muvm) + `libkrun` from (git clone -b fixkrun https://github.com/dorindabassey/libkrun) + latest rutabaga_gfx(this PR) by running `xeyes` in the vm.

dorindabassey commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -19,13 +19,22 @@ pub const CROSS_DOMAIN_CMD_SEND: u8 = 4;
+/// The maximum number of identifiers (inspired by sommelier, protocol v0)
 pub const CROSS_DOMAIN_MAX_IDENTIFIERS: usize = 28;
+/// As above, but for newer protocol (value inspired by wp_linux_dmabuf, protocol v1+)
+pub const CROSS_DOMAIN_MAX_IDENTIFIERS_V2: usize = 4;
```
Thanks, somehow i missed that change but fixed now. I hope it's ok I merged it in the 3rd commit?

dorindabassey commented on Cargo.toml:
```diff
@@ -30,6 +30,7 @@ vulkano = ["dep:vulkano"]
 cfg-if = "1.0.0"
 libc = "0.2.153"
 log = "0.4"
+nix = { version = "0.29", features = ["mman"] }
```
Right, Thanks!

gurchetansingh commented on src/rutabaga_utils.rs:
```diff
@@ -608,6 +608,8 @@ impl Transfer3D {
 /// Rutabaga path types
 pub const RUTABAGA_PATH_TYPE_WAYLAND: u32 = 0x0001;
 pub const RUTABAGA_PATH_TYPE_GPU: u32 = 0x0002;
+pub const RUTABAGA_PATH_TYPE_PW: u32 = 0x0010;
```
nit: spell out the entire path `RUTABAGA_PATH_TYPE_PIPEWIRE`

gurchetansingh commented on src/rutabaga_core.rs:
```diff
@@ -63,6 +64,8 @@ use crate::snapshot::RutabagaSnapshotWriter;
 use crate::virgl_renderer::VirglRenderer;
 use crate::RutabagaPaths;
 
+pub type VirtioFsTable = Arc<Mutex<Map<(u64, u64), File>>>;
```
Can you define what both of these pairs mean?

` VirtioFsKey {
     <descriptor_name_1>: u64,
     <descriptive_name_2>: u64,
}`
Arc<Mutex<Map<VirtioFsKey, File>>>

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -120,3 +215,37 @@ pub struct CrossDomainReadWrite {
+#[derive(Copy, Clone, Default, FromBytes, IntoBytes, Immutable)]
+pub struct CrossDomainFutexNew {
+    pub hdr: CrossDomainHeader,
+    pub fs_id: u64,
```
document the difference between `fs_id`, `handle`, and `id`

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -449,6 +571,109 @@ impl CrossDomainWorker {
+                    let event = Event::try_from(MesaHandle {
+                        os_handle: owned_desc,
+                        handle_type: MESA_HANDLE_TYPE_SIGNAL_EVENT_FD,
+                    })?;
```
should we be using descriptor.determine_type()

dorindabassey replied:
I changed it to use descriptor.determine_type() but i guess I'll have to upstream this to mesa aswell, is that right?

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -120,3 +128,29 @@ pub struct CrossDomainReadWrite {
+
+#[repr(C)]
+#[derive(Copy, Clone, Default, FromBytes, IntoBytes, Immutable)]
+pub struct CrossDomainFutexNew {
```
futex is very linux specific.  maybe we call it what the kernel/OS actually does?  `CrossDomainMemoryWatcher`??

dorindabassey replied:
Ok, changing all "futex" related terminologies to "MemoryWatcher"

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -38,6 +42,8 @@ pub const CROSS_DOMAIN_ID_TYPE_READ_PIPE: u32 = 3;
 /// The host receives the write end of the pipe over the host Wayland socket.
 pub const CROSS_DOMAIN_ID_TYPE_WRITE_PIPE: u32 = 4;
 
+pub const CROSS_DOMAIN_ID_TYPE_SHM: u32 = 5;
```
`CROSS_DOMAIN_ID_TYPE_SHM` not used?

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -19,10 +19,14 @@ pub const CROSS_DOMAIN_CMD_SEND: u8 = 4;
 pub const CROSS_DOMAIN_CMD_RECEIVE: u8 = 5;
 pub const CROSS_DOMAIN_CMD_READ: u8 = 6;
 pub const CROSS_DOMAIN_CMD_WRITE: u8 = 7;
+pub const CROSS_DOMAIN_CMD_FUTEX_NEW: u8 = 8;
```
`CROSS_DOMAIN_CMD_CREATE_MEMORY_WATCHER`

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -19,10 +19,14 @@ pub const CROSS_DOMAIN_CMD_SEND: u8 = 4;
 pub const CROSS_DOMAIN_CMD_READ: u8 = 6;
 pub const CROSS_DOMAIN_CMD_WRITE: u8 = 7;
+pub const CROSS_DOMAIN_CMD_FUTEX_NEW: u8 = 8;
+pub const CROSS_DOMAIN_CMD_FUTEX_SIGNAL: u8 = 9;
```
`CROSS_DOMAIN_CMD_SIGNAL_MEMORY_WATCHER`

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -19,10 +19,14 @@ pub const CROSS_DOMAIN_CMD_SEND: u8 = 4;
 pub const CROSS_DOMAIN_CMD_WRITE: u8 = 7;
+pub const CROSS_DOMAIN_CMD_FUTEX_NEW: u8 = 8;
+pub const CROSS_DOMAIN_CMD_FUTEX_SIGNAL: u8 = 9;
+pub const CROSS_DOMAIN_CMD_FUTEX_DESTROY: u8 = 10;
```
`CROSS_DOMAIN_CMD_DESTROY_MEMORY_WATCHER`

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -47,6 +53,8 @@ pub const CROSS_DOMAIN_CHANNEL_RING: u32 = 1;
 /// Read pipe IDs start at this value.
 pub const CROSS_DOMAIN_PIPE_READ_START: u32 = 0x80000000;
+/// Futex IDs start at this value.
+pub const CROSS_DOMAIN_FUTEX_START: u32 = 0x40000000;
```
`MEMORY_WATCHER_START``

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -5,18 +5,28 @@
 
+#[cfg(target_os = "linux")]
+use log::{error, info};
+use rustix::mm::{mmap, munmap, MapFlags, ProtFlags};
```
Can we wrap the futex mapping in a `MemoryMapping` in `mesa3d_util`?

dorindabassey replied:
yes we can use MemoryMapping from mesa3d_util instead.

gurchetansingh reviewed (commented):
Sorry for the late response (holiday season).   Needs to pass the CI, and try to make it more OS-agnostic: i.e, less (`target_os=linux`)

CC: @valpackett for any other review comments.

valpackett commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -19,12 +19,19 @@ pub const CROSS_DOMAIN_CMD_SEND: u8 = 4;
+pub const CROSS_DOMAIN_CHANNEL_TYPE_X11: u32 = 0x0011;
 
-/// The maximum number of identifiers
+/// The maximum number of identifiers (must match what sommelier expects)
```
nit: better to drop this comment change. sommelier is kinda dead; rutabaga is what defines the API

valpackett commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -109,6 +153,57 @@ pub struct CrossDomainSendReceive {
+
+#[repr(C)]
+#[derive(Copy, Clone, Default, FromBytes, IntoBytes, Immutable)]
+pub struct CrossDomainSendReceiveV2 {
```
If there's no change in `MAX_IDENTIFIERS`.. this V2 is completely identical to the "V1", isn't it? We can drop this whole `CrossDomainSendReceiveBase` thing.

dorindabassey replied:
Right! Dropping this.

valpackett commented on src/cross_domain/mod.rs:
```diff
@@ -5,22 +5,36 @@
+#[cfg(target_os = "linux")]
+use std::sync::atomic::AtomicBool;
+#[cfg(target_os = "linux")]
+use std::sync::atomic::{AtomicU32, Ordering};
```
```suggestion
#[cfg(target_os = "linux")]
use std::sync::atomic::{AtomicBool, AtomicU32, Ordering};
```

valpackett commented on src/cross_domain/mod.rs:
```diff
@@ -115,6 +149,96 @@ struct CrossDomainWorker {
+#[cfg(target_os = "linux")]
+struct CrossDomainFutex {
+    address: FutexPtr,
+    #[allow(dead_code)] // Kept alive for RAII, dropped with CrossDomainFutex
```
`_underscore` prefix the field name instead?

valpackett commented on src/cross_domain/mod.rs:
```diff
@@ -449,6 +571,109 @@ impl CrossDomainWorker {
+            if let Ok(seek_size) = f.seek(SeekFrom::End(0)) {
+                *identifier_type = CROSS_DOMAIN_ID_TYPE_VIRTGPU_BLOB;
+                *identifier_size = seek_size.try_into().map_err(MesaError::TryFromIntError)?;
+                let fd_path = read_link(format!("/proc/self/fd/{}", raw_fd))
```
`/proc/self/fd` is a Linux convention, there should probably be a `cfg()` soooomewhere around here?

gurchetansingh commented on subprojects/packagefiles/rustix-1-rs/meson.build:
```diff
@@ -54,11 +56,11 @@ lib = static_library(
 
 dep_rustix = declare_dependency(
-  link_with : [lib, errno, libc, bitflags] + os_deps,
+  link_with : [lib, errno, libc, bitflags, linux_raw_sys] + os_deps,
```
The Android Rust team unfortunately n'acked `linux_raw_sys` back-end for `rustix` (https://gitlab.freedesktop.org/mesa/mesa/-/issues/13378), so we probably want to avoid it for portability reasons.  Though, I'm not sure why you need it?

Rustix has futex in `./src/backend/libc/thread/futex.rs`.

dorindabassey replied:
Thanks, removed it.

gurchetansingh commented on third_party/mesa3d/src/util/rust/lib.rs:
```diff
@@ -25,6 +25,7 @@ pub use sys::platform::descriptor::OwnedDescriptor;
 pub use sys::platform::descriptor::RawDescriptor;
 pub use sys::platform::descriptor::DEFAULT_RAW_DESCRIPTOR;
 pub use sys::platform::event::Event;
+pub use sys::platform::futex;
```
Rename this to `os_memory_watcher `.  Only in the Linux backend should we use the `futex` nomenclature.

gurchetansingh commented on src/rutabaga_core.rs:
```diff
@@ -63,6 +64,17 @@ use crate::snapshot::RutabagaSnapshotWriter;
+    /// VirtioFS filesystem ID (identifies which virtio-fs instance)
+    pub fs_id: u64,
+    /// File handle within the filesystem (identifies a specific file)
+    pub handle: u64,
```
nice!

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -120,3 +153,40 @@ pub struct CrossDomainReadWrite {
+
+#[repr(C)]
+#[derive(Copy, Clone, Default, FromBytes, IntoBytes, Immutable)]
+pub struct CrossDomainMemoryWatcherNew {
```
If you have `CROSS_DOMAIN_CMD_CREATE_ATOMIC_MEMORY_SENTINEL`, might as well call this `CrossDomainCreateAtomicMemorySentinel`

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -120,3 +153,40 @@ pub struct CrossDomainReadWrite {
+
+#[repr(C)]
+#[derive(Copy, Clone, Default, FromBytes, IntoBytes, Immutable)]
+pub struct CrossDomainMemoryWatcherSignal {
```
`CrossDomainSignalAtomicMemorySentinel`

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -120,3 +153,40 @@ pub struct CrossDomainReadWrite {
+
+#[repr(C)]
+#[derive(Copy, Clone, Default, FromBytes, IntoBytes, Immutable)]
+pub struct CrossDomainMemoryWatcherDestroy {
```
`CrossDomainDestroyAtomicMemorySentinel`

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -120,3 +153,40 @@ pub struct CrossDomainReadWrite {
+
+#[repr(C)]
+#[derive(Copy, Clone, Default, FromBytes, IntoBytes, Immutable)]
+pub struct CrossDomainReadEventfdNew {
```
`CrossDomainCreateEvent`

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -19,10 +19,17 @@ pub const CROSS_DOMAIN_CMD_SEND: u8 = 4;
+pub const CROSS_DOMAIN_CMD_CREATE_MEMORY_WATCHER: u8 = 8;
+pub const CROSS_DOMAIN_CMD_SIGNAL_MEMORY_WATCHER: u8 = 9;
+pub const CROSS_DOMAIN_CMD_DESTROY_MEMORY_WATCHER: u8 = 10;
+pub const CROSS_DOMAIN_CMD_READ_EVENTFD_NEW: u8 = 11;
```
`CROSS_DOMAIN_CMD_READ_CREATE_EVENT`

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -19,10 +19,17 @@ pub const CROSS_DOMAIN_CMD_SEND: u8 = 4;
+pub const CROSS_DOMAIN_CMD_SIGNAL_MEMORY_WATCHER: u8 = 9;
+pub const CROSS_DOMAIN_CMD_DESTROY_MEMORY_WATCHER: u8 = 10;
+pub const CROSS_DOMAIN_CMD_READ_EVENTFD_NEW: u8 = 11;
+pub const CROSS_DOMAIN_CMD_READ_EVENTFD_DESTROY: u8 = 12;
```
`CROSS_DOMAIN_CMD_READ_DESTROY_EVENT`

But I don't see this being used in the code?

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -38,6 +45,8 @@ pub const CROSS_DOMAIN_ID_TYPE_READ_PIPE: u32 = 3;
 /// The host receives the write end of the pipe over the host Wayland socket.
 pub const CROSS_DOMAIN_ID_TYPE_WRITE_PIPE: u32 = 4;
 
+pub const CROSS_DOMAIN_ID_TYPE_EVENTFD: u32 = 6;
```
`CROSS_DOMAIN_ID_TYPE_EVENT`.  Why 6 instead of the already available 5?

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -5,22 +5,45 @@
-use mesa3d_util::AsRawDescriptor;
-use mesa3d_util::DescriptorType;
 use mesa3d_util::Event;
+#[cfg(any(target_os = "android", target_os = "linux"))]
```
Don't need OS specfic guards.  MappedRegion, MemoryMapping, FromRawDescriptor, IntoRawDescriptor all OS agnostic.

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -5,22 +5,45 @@
+#[cfg(unix)]
+use std::os::unix::io::{FromRawFd, IntoRawFd};
+#[cfg(any(target_os = "android", target_os = "linux"))]
+use std::sync::atomic::{AtomicBool, AtomicU32, Ordering};
```
`std::sync::atomic::` compiles are all target platforms (Linux, Android, Windows, MacOS).  But should be the `atomic_memory_sentinel.rs` file.

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -117,16 +162,106 @@ struct CrossDomainWorker {
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+impl CrossDomainMemoryWatcher {
+    fn watcher_thread(
```
Usually when you have a structure that encapsulates a thread, there are often two functions:

`CrossDomainAtmicMemorySentinelThread::new` +
 `CrossDomainAtmicMemorySentinelThread::run` (what is no `watcher_thread` function)

I suggest you use that idiom here.

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -117,16 +162,106 @@ struct CrossDomainWorker {
+#[cfg(any(target_os = "android", target_os = "linux"))]
+unsafe impl Send for FutexPtr {}
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
```
If we use OS-agnostic abstractions with the `CrossDomainAtomicMemorySentinelWorker`, we can eliminate `#[cfg(any(target_os = "android", target_os = "linux"))]` in `src/cross_domain`.  The OS-specific code should be in the backends of `mesa3d_util`.

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -328,63 +466,7 @@ impl CrossDomainWorker {
         if let Some(event) = events.first() {
             match event.connection_id {
                 CROSS_DOMAIN_CONTEXT_CHANNEL_ID => {
-                    let (len, files) = self.state.receive_msg(receive_buf)?;
```
maybe split this into:

1) one prep commit that moves recieve to it's own function
2) the X11/pipewire commit which adds changes on that.  I feel most of the change is recieving `Events` now?

dorindabassey replied:
done

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -403,21 +485,61 @@ impl CrossDomainWorker {
+                id if id >= CROSS_DOMAIN_MEMORY_WATCHER_START as u64
+                    && id < CROSS_DOMAIN_PIPE_READ_START as u64 =>
+                {
+                    #[cfg(any(target_os = "android", target_os = "linux"))]
```
shouldn't need the cfg

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -434,11 +556,19 @@ impl CrossDomainWorker {
                             }
                         }
+                        #[cfg(any(target_os = "android", target_os = "linux"))]
+                        CrossDomainItem::Eventfd(ref efd) => {
```
`CrossDomainItem::Event`

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -434,11 +556,19 @@ impl CrossDomainWorker {
                                 self.wait_ctx.delete(readpipe.as_borrowed_descriptor())?;
                             }
                         }
+                        #[cfg(any(target_os = "android", target_os = "linux"))]
```
delete

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -47,6 +56,8 @@ pub const CROSS_DOMAIN_CHANNEL_RING: u32 = 1;
 /// Read pipe IDs start at this value.
 pub const CROSS_DOMAIN_PIPE_READ_START: u32 = 0x80000000;
+/// Memory watcher IDs start at this value.
+pub const CROSS_DOMAIN_MEMORY_WATCHER_START: u32 = 0x40000000;
```
CROSS_DOMAIN_ATOMIC_MEMORY_SENTINEL_START

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -74,17 +102,32 @@ const CROSS_DOMAIN_DEFAULT_BUFFER_SIZE: usize = 4096;
+type Sender = Event;
+type Receiver = Event;
+
+fn channel_signal(sender: &mut Sender) -> RutabagaResult<()> {
```
Delete all instances of channel_signal -- it's an holdover from a prior error.  You can just call `Event::Signal`

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -74,17 +102,32 @@ const CROSS_DOMAIN_DEFAULT_BUFFER_SIZE: usize = 4096;
+    #[allow(dead_code)] // `WaylandReadPipe` is never constructed on Windows.
     WaylandReadPipe(ReadPipe),
     WaylandWritePipe(WritePipe),
+    #[cfg(any(target_os = "android", target_os = "linux"))]
```
`Event` compiles fine on other platforms

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -74,17 +102,32 @@ const CROSS_DOMAIN_DEFAULT_BUFFER_SIZE: usize = 4096;
 enum CrossDomainItem {
     ImageRequirements(ImageMemoryRequirements),
     Blob(MesaHandle),
+    #[allow(dead_code)] // `WaylandReadPipe` is never constructed on Windows.
```
Delete: Wayland readpipe does construct fine on Windows

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -117,16 +162,106 @@ struct CrossDomainWorker {
+    watcher_thread: Option<thread::JoinHandle<()>>,
+    shutdown: Arc<AtomicBool>,
+    evt: Arc<Receiver>,
+}
```
The `CrossDomainMemoryWatcher` structure needs to be refactored. 

Maybe a  `CrossDomainAtomicMemorySentinelWorker` should contain `Arc<Mutex<AtomicMemorySentinel>>`  and Events only (atomic book for shutdown also seems good)?

dorindabassey replied:
IIUC, do you mean the AtomicBool for shutdown should be inside `AtomicMemorySentinel`, and `CrossDomainAtomicMemorySentinelWorker` should only contain `Arc<Mutex<AtomicMemorySentinel>>` and Arc<Event>?

gurchetansingh replied:
`CrossDomainAtomicMemorySentinelWorker` should contain the synchronization primitives (`AtomicBool`, `Events` etc.) needed to synchronize with other threads.

Thinking about it even more, `CrossDomainAtomicMemorySentinelWorker` should contain `AtomicMemorySentinel`, and you can make `AtomicMemorySentinel` safe to share across threads (it is an atomic, and `MemoryMapping` already has this property).

dorindabassey replied:
Thanks! Fixed it, can you confirm if this is ok? I added the new AtomicMemorySentinel and made it shareable,  so now it's just a pure atomic memory abstraction with 
```
pub struct AtomicMemorySentinel {
    #[allow(dead_code)] // Kept alive for RAII
    memory_mapping: MemoryMapping,
    atomic_ptr: AtomicPtr,
}

gurchetansingh reviewed (changes requested):
Okay, finally wrapping my head around what this is trying to do and have more substantive feedback.  I think we should create a:

`third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs` file.

and that would be the access point to `futex` and `WaitOnAddress` like APIs.  

```
use crate::sys::platform::futex as futex

AtomicMemorySentinel {
     memory_mapping: MemoryMapping,
     atomic_ptr: *mut AtomicU32,
     value: ..
}

```
impl Drop for AtomicMemorySentinel {
    fn drop(&mut self) {
        futex::wake_all(...)
    }
}
```
```
impl AtomicMemorySentinel {
     pub fn new(mapping: MemoryMapping) -> MesaResult<AtomicMemorySentinel>
     pub fn wait() -> MesaResult<()>;
     pub fn signal() -> MesaResult<()>
```
}
```

And then in the CrossDomain, code you can work in terms of `Arc<Mutex<AtomicMemorySentinel>`.  I feel like the current `CrossDomainMemoryWatcher` is mixing thread operations, atomic memory operations.  

Plus, `AtomicMemorySentinel` sounds coller than `MemoryWatcher`, does it not?!?

Also, `cross_domain.rs` is getting rather large.  Feel free to split into seperate files: I.e, `cross_domain_worker.rs`, `cross_domain_atomic_memory_sentinel_worker.rs`.  Thanks again for your patience!

Commit:
cross-domain: Extract process_receive and add Event support

Extract file descriptor receive logic from handle_fence()
into a separate process_receive() method to improve code
organization.

Co-developed-by: Sasha Finkelstein <fnkl.kernel@gmail.com>
Signed-off-by: Dorinda Bassey <dbassey@redhat.com>

dorindabassey commented:
> Okay, finally wrapping my head around what this is trying to do and have more substantive feedback. I think we should create a:
> 
> `third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs` file.
> 
> and that would be the access point to `futex` and `WaitOnAddress` like APIs.
> 
> ```
> use crate::sys::platform::futex as futex
> 
> AtomicMemorySentinel {
>      memory_mapping: MemoryMapping,
>      atomic_ptr: *mut AtomicU32,
>      value: ..
> }
> ```
> 
> impl Drop for AtomicMemorySentinel { fn drop(&mut self) { futex::wake_all(...) } }
> 
> impl AtomicMemorySentinel { pub fn new(mapping: MemoryMapping) -> MesaResult pub fn wait() -> MesaResult<()>; pub fn signal() -> MesaResult<()>
> 
> ```
> }
> ```
> 
> And then in the CrossDomain, code you can work in terms of `Arc<Mutex<AtomicMemorySentinel>`. I feel like the current `CrossDomainMemoryWatcher` is mixing thread operations, atomic memory operations.
> 
> Plus, `AtomicMemorySentinel` sounds coller than `MemoryWatcher`, does it not?!?
> 
> Also, `cross_domain.rs` is getting rather large. Feel free to split into seperate files: I.e, `cross_domain_worker.rs`, `cross_domain_atomic_memory_sentinel_worker.rs`. Thanks again for your patience!

Thanks for the detailed feedback! I've addressed all the other points. Regarding splitting cross_domain/mod.rs into separate files (cross_domain_worker.rs, cross_domain_atomic_memory_sentinel_worker.rs): I'd prefer to defer this to a follow-up change since I need to rebase this series first, and the file split would make the rebase messy.
Also I don't know what the last CI error is about. I thought I removed `linux_raw_sys`

gurchetansingh commented on Cargo.toml:
```diff
@@ -37,6 +37,9 @@ serde = { version = "1.0", features = ["derive"] }
 mesa3d_util = { path = "third_party/mesa3d/src/util/rust/", version = "0.1.75" }
 
+[target.'cfg(any(target_os = "android", target_os = "linux"))'.dependencies]
+rustix = { version = "1.0.7", features = ["mm"] }
```
You no long Rustix as a direct dependency.  You can remove this.

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -19,10 +19,16 @@ pub const CROSS_DOMAIN_CMD_SEND: u8 = 4;
 pub const CROSS_DOMAIN_CMD_WRITE: u8 = 7;
+pub const CROSS_DOMAIN_CMD_CREATE_MEMORY_WATCHER: u8 = 8;
+pub const CROSS_DOMAIN_CMD_SIGNAL_MEMORY_WATCHER: u8 = 9;
+pub const CROSS_DOMAIN_CMD_DESTROY_MEMORY_WATCHER: u8 = 10;
```
These should be `CROSS_DOMAIN_CMD_*_ATOMIC_MEMORY_SENTINEL` for consistency with `CROSS_DOMAIN_ATOMIC_MEMORY_SENTINEL_START`.

gurchetansingh commented on src/cross_domain/cross_domain_protocol.rs:
```diff
@@ -80,14 +90,36 @@ pub struct CrossDomainHeader {
+    pub query_ring_id: u32,
+    pub channel_ring_id: u32,
+    pub channel_type: u32,
+    pub protocol_version: u32,
```
It doesn't seem the code is actually `protocol_version` right now?  I see the rationale: PING_PROTOCOL version is a common idiom.  But can we avoid in this CL until we have use case?

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -32,6 +36,8 @@ use mesa3d_util::WaitTimeout;
 use mesa3d_util::WritePipe;
 use mesa3d_util::MESA_HANDLE_TYPE_MEM_DMABUF;
 use mesa3d_util::MESA_HANDLE_TYPE_MEM_SHM;
+#[cfg(any(target_os = "android", target_os = "linux"))]
```
you can drop this: this compiles everywhere

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -74,17 +83,25 @@ const CROSS_DOMAIN_DEFAULT_BUFFER_SIZE: usize = 4096;
 const CROSS_DOMAIN_MAX_SEND_RECV_SIZE: usize =
     CROSS_DOMAIN_DEFAULT_BUFFER_SIZE - size_of::<CrossDomainSendReceive>();
 
+// Type aliases for event signaling
```
Remove the `Reciever` alias

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -74,17 +83,25 @@ const CROSS_DOMAIN_DEFAULT_BUFFER_SIZE: usize = 4096;
     HandleFence(RutabagaFence),
     AddReadPipe(u32),
     Finish,
+    #[cfg(any(target_os = "android", target_os = "linux"))]
```
Remove this: this should compile everywhere

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -74,17 +83,25 @@ const CROSS_DOMAIN_DEFAULT_BUFFER_SIZE: usize = 4096;
+    #[cfg(any(target_os = "android", target_os = "linux"))]
+    AddMemoryWatcher(u32, Arc<Receiver>),
+    #[cfg(any(target_os = "android", target_os = "linux"))]
+    AddReadEventfd(u32),
```
`AddReadEvent`.  Remove the OS-specific conditionals too.

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -117,16 +136,50 @@ struct CrossDomainWorker {
+    sentinel: Arc<AtomicMemorySentinel>,
+    watcher_thread: Option<thread::JoinHandle<()>>,
+    shutdown: Arc<AtomicBool>,
+    evt: Arc<Receiver>,
```
`evt: Arc<Receiver>` is never used in  `CrossDomainAtomicMemorySentinelWorker`.  I think you should remove, and that should allow you to remove the `Arc<Reciever>` a.k.a `Arc<Event>` elsewhere, and just use  `Event`.

gurchetansingh commented on third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs:
```diff
@@ -0,0 +1,133 @@
+/// Safe to share across threads.
+pub struct AtomicMemorySentinel {
+    #[allow(dead_code)] // Kept alive for RAII
+    memory_mapping: MemoryMapping,
```
you can just do `_memory_mapping` to avoid the warning

gurchetansingh commented on third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs:
```diff
@@ -0,0 +1,133 @@
+unsafe impl Send for AtomicMemorySentinel {}
+unsafe impl Sync for AtomicMemorySentinel {}
+
+/// Thread worker for monitoring atomic memory changes.
```
`AtomicMemorySentinelThread` should be `cross_domain/atomic_memory_sentinel_manager.rs`

gurchetansingh commented on subprojects/packagefiles/rustix-1-rs/meson.build:
```diff
@@ -29,6 +29,7 @@ if host_machine.system() == 'linux' or host_machine.system() == 'android'
   '--cfg', 'feature="net"',
   '--cfg', 'feature="param"',
   '--cfg', 'feature="pipe"',
+  '--cfg', 'feature="thread"',
```
hmm, this will introduce a `linux-raw-sys` dependency in Android, which the Rust team will not be happy about.   I say keep this, and we'll have to get another patch through rustix before merged, but don't add a `linux-raw-sys` dep in the meson builds.

gurchetansingh reviewed (commented):
While you don't need to refactor existing code, I still you should encapsulate the new functionality into it's own file to the extent possibility, since it does introduce complex spread throughout the place.

I suggest the creation of `atomic_memory_sentinel_manager.rs` in the `src/cross_domain` dir.  That can encapsulate management of the `AtomicMemorySentinelThread` (s) and access to them.  You don't have prefix everything with `CrossDomain` in `src/cross_domain.rs` -- I think we'll refactor that pattern later.    

mod.rs
```
 struct CrossDomainContext {
     watcher_manager: Arc<Mutex<AtomicMemorySentinelManager>>,
     worker_thread: Option<thread::JoinHandle<()>>,
 }
 
 impl CrossDomainContext {
     fn memory_watcher_new(&mut self, cmd: &CreateNewWatcherCmd) -> Result<()> {
         // Lock the mutex to get mutable access to the manager
         let mut manager = self.watcher_manager.lock().unwrap();
         let (id, eventfd_descriptor) = manager.create_watcher(cmd.fs_id, cmd.handle)?;
 
         self.state.add_job(CrossDomainJob::AddPolledDescriptor(id, eventfd_descriptor));
         Ok(())
     }
 
     fn memory_watcher_destroy(&mut self, cmd: &DestroyWatcherCmd) -> Result<()> {
         let mut manager = self.watcher_manager.lock().unwrap();
         manager.destroy_watcher(cmd.id)?;
 
         self.state.add_job(CrossDomainJob::RemovePolledDescriptor(cmd.id));
         Ok(())
     }
 }
 
 
 struct CrossDomainWorker {
     watcher_manager: Arc<Mutex<AtomicMemorySentinelManager>>,
     wait_ctx: WaitContext,
 }
 
 impl CrossDomainWorker {
     fn handle_fence(&mut self) -> Result<()> {
         let events = self.wait_ctx.wait()?;
         if let Some(event) = events.first() {
             if event.id >= MEMORY_WATCHER_ID_START {
                 let manager = self.watcher_manager.lock().unwrap();
 
                 let ring_message = manager.handle_event(event.id, &event.descriptor)?;
 
                 self.state.write_to_ring(RingWrite::Write(ring_message, None), ...)?;
             }
         }
         Ok(())
}

atomic_memory_sentinel_manager.rs
 pub struct AtomicMemorySentinelManager {
     watchers: Map<u32, SentinenlInstance>,
     next_id: u32,
     virtiofs_table: VirtioFsTable, // A reference to the required dependency
 }
 
 impl AtomicMemorySentinelManager {
     pub fn new(virtiofs_table: VirtioFsTable) -> Self {
         Self {
             watchers: Map::new(),
             next_id: MEMORY_WATCHER_ID_START,
             virtiofs_table,
         }
     }
 
     pub fn create_watcher(&mut self, fs_id: u64, handle: u64) -> Result<(u32, OwnedDescriptor)> {
         let eventfd = Event::new()?;
         let thread_eventfd = eventfd.try_clone()?;
 
         let join_handle = thread::spawn(move || {
             // Thread logic that waits on the sentinel and signals `thread_eventfd`.
         });
 
         Ok((id, eventfd.into()))
     }
 
     pub fn handle_event(&self, id: u32, event_fd: &Event) -> Result<RingMessage> {
        .. 
         Ok(message)
     }
 
     pub fn destroy_watcher(&mut self, id: u32) -> Result<()> {
        ...
     }
 }
 
 impl Drop for AtomicMemorySentinelManager {
     fn drop(&mut self) {
         for (_, watcher) in self.watchers.drain() {
             watcher.shutdown();
         }
     }
 }
```

gurchetansingh commented on src/cross_domain/atomic_memory_sentinel_manager.rs:
```diff
@@ -0,0 +1,273 @@
+            info!("Atomic memory sentinel worker dropped without shutdown");
+            self.shutdown();
+        }
+        // MemoryMapping automatically unmaps in its own Drop
```
MemoryMapping is no longer revelant.

gurchetansingh commented on third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs:
```diff
@@ -0,0 +1,75 @@
+// Wrapper to make the pointer Send + Sync
+struct AtomicPtr(*mut AtomicU32);
+unsafe impl Send for AtomicPtr {}
+unsafe impl Sync for AtomicPtr {}
```
I think we can use `std::sync::atomic::AtomicPtr`

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -434,11 +424,18 @@ impl CrossDomainWorker {
                         }
+                        CrossDomainItem::Event(ref efd) => {
+                            // For eventfd, we can use wait() to consume the event
+                            if event.readable {
```
`ref event`

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -449,6 +446,96 @@ impl CrossDomainWorker {
+
+        for ((identifier, identifier_type, identifier_size), file) in iter {
+            // Platform-specific file descriptor handling
+            #[cfg(any(target_os = "android", target_os = "linux"))]
```
remove OS conditional

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -746,6 +855,65 @@ impl CrossDomainContext {
         Ok(())
     }
 
+    #[cfg(any(target_os = "android", target_os = "linux"))]
```
delete all os conditionals here (including for destroy, new, signal)

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -746,6 +855,65 @@ impl CrossDomainContext {
+    }
+
+    #[cfg(any(target_os = "android", target_os = "linux"))]
+    fn read_eventfd_new(&mut self, cmd_eventfd_new: &CrossDomainCreateEvent) -> RutabagaResult<()> {
```
`read_event_new`: not `read_eventfd_new` (we're very big on the cross-platform here).

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -79,12 +84,15 @@ enum CrossDomainItem {
     HandleFence(RutabagaFence),
     AddReadPipe(u32),
     Finish,
+    AddMemoryWatcher(u32, Event),
```
`AddAtomicMemorySentinel`

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -94,6 +102,7 @@ enum RingWrite<'a, T> {
 
 type CrossDomainJobs = Mutex<Option<VecDeque<CrossDomainJob>>>;
 type CrossDomainItemState = Arc<Mutex<CrossDomainItems>>;
+type WatcherManager = Arc<Mutex<AtomicMemorySentinelManager>>;
```
`SentinelManager`

gurchetansingh commented on src/cross_domain/atomic_memory_sentinel_manager.rs:
```diff
@@ -0,0 +1,273 @@
+        let handle = OwnedDescriptor::from(file);
+        let mapping = MemoryMapping::from_safe_descriptor(
+            handle,
+            4,
```
`std::mem::size_of::<u32>`

gurchetansingh commented on third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs:
```diff
@@ -0,0 +1,75 @@
+
+use std::sync::atomic::{AtomicU32, Ordering};
+
+use crate::sys::platform::os_memory_watcher;
```
`crate::sys::platform::atomic_memory_sentinel as platform_atomic_memory_sentinel`

gurchetansingh commented on third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs:
```diff
@@ -0,0 +1,75 @@
+impl AtomicMemorySentinel {
+    /// Create a new AtomicMemorySentinel from a memory mapping.
+    pub fn new(memory_mapping: MemoryMapping) -> MesaResult<Self> {
+        let atomic_ptr = AtomicPtr(memory_mapping.as_ptr() as *mut AtomicU32);
```
check the memory_mapping.size() before doing this cast

gurchetansingh commented on third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs:
```diff
@@ -0,0 +1,79 @@
+    /// This will wake up waiters on this sentinel.
+    pub fn signal(&self) -> MesaResult<()> {
+        let atomic_val = unsafe { &*self.atomic_ptr.load(Ordering::Relaxed) };
+        platform_os_memory_watcher::wake_bitset(atomic_val, i32::MAX, 1);
```
nit[consistent naming]: `platform_atomic_memory_sentinel` instead `platform_os_memory_watcher`

gurchetansingh commented on third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs:
```diff
@@ -0,0 +1,79 @@
+use std::mem::size_of;
+use std::sync::atomic::{AtomicPtr, AtomicU32, Ordering};
+
+use crate::sys::platform::os_memory_watcher as platform_os_memory_watcher;
```
nit[consistent naming]: `crate::sys::platform::atomic_memory_sentinel`  as `atomic_memory_sentinel`

gurchetansingh commented on third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs:
```diff
@@ -0,0 +1,79 @@
+
+    /// Wake all threads waiting on this sentinel.
+    pub fn wake_all(&self) {
+        let atomic_val = unsafe { &*self.atomic_ptr.load(Ordering::Relaxed) };
```
Add safety comment

gurchetansingh commented on third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs:
```diff
@@ -0,0 +1,79 @@
+
+    /// Store a value to atomic memory.
+    pub fn store(&self, val: u32) {
+        let atomic_val = unsafe { &*self.atomic_ptr.load(Ordering::Relaxed) };
```
safety comment

gurchetansingh commented on third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs:
```diff
@@ -0,0 +1,79 @@
+}
+
+unsafe impl Send for AtomicMemorySentinel {}
+unsafe impl Sync for AtomicMemorySentinel {}
```
Safety comment.

gurchetansingh commented on third_party/mesa3d/src/util/rust/lib.rs:
```diff
@@ -25,6 +27,7 @@ pub use sys::platform::descriptor::OwnedDescriptor;
 pub use sys::platform::descriptor::RawDescriptor;
 pub use sys::platform::descriptor::DEFAULT_RAW_DESCRIPTOR;
 pub use sys::platform::event::Event;
+pub use sys::platform::os_memory_watcher;
```
don't make this a public use

gurchetansingh commented on third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs:
```diff
@@ -0,0 +1,79 @@
+/// Safe to share across threads.
+pub struct AtomicMemorySentinel {
+    _memory_mapping: MemoryMapping,
+    atomic_ptr: AtomicPtr<AtomicU32>,
```
Oops, I think we somehow introduced double-atomicity.  

Every time you want to use the sentinel, you perform an atomic load to get the pointer, then another atomic operation on the u32 value it points to.  The first load is redundant.

I think `AtomicU32::from_ptr(..) is what we need.  Using it safely requires these conditions:

https://doc.rust-lang.org/stable/std/sync/atomic/struct.AtomicPtr.html#safety

dorindabassey replied:
Fixed, Thanks!

gurchetansingh commented on src/cross_domain/atomic_memory_sentinel_manager.rs:
```diff
@@ -0,0 +1,273 @@
+
+use log::info;
+use mesa3d_util::os_memory_watcher as platform_os_memory_watcher;
+use mesa3d_util::AtomicMemorySentinel;
```
delete `mesa3d_util::os_memory_watcher as platform_os_memory_watcher`.  We should leak platform-specific code here.

gurchetansingh commented on src/cross_domain/atomic_memory_sentinel_manager.rs:
```diff
@@ -0,0 +1,273 @@
+
+        loop {
+            // This returns when the futex is woken up OR if the value has changed.
+            platform_os_memory_watcher::wait_bitset(atomic_val, val, 1);
```
* Extend `AtomicMemorySentinel`: Add a wait method to AtomicMemorySentinel in third_party/mesa3d/src/util/rust/atomic_memory_sentinel.rs.

```
/// Blocks until the value changes from `val` or is woken up.                                                                                     │
│pub fn wait(&self, val: u32) {
```

`AtomicMemorySentinel` should encapsulate the atomic value and avoid leaking references to it.  This implies `atomic_val()` should be deleted.

gurchetansingh commented on src/cross_domain/mod.rs:
```diff
@@ -746,6 +838,61 @@ impl CrossDomainContext {
         Ok(())
     }
 
+    fn memory_watcher_signal(
```
[nit]: consistent naming `atomic_memory_sentinel_signal` + `manager.signal_atomic_memory_sentinel` + `cmd_atomic_memory_sentinel_signal`.  Nit is repeated for `memory_watcher_destroy` and `memory_watcher_destroy`

gurchetansingh reviewed (changes requested):
Mostly nits, but I think we need to get rid of the double-atomicity and leaking a reference to the atomic value.

I'll also prepare a CL to solve the Rustix/Android issue..

dorindabassey commented:
> Mostly nits, but I think we need to get rid of the double-atomicity and leaking a reference to the atomic value.
> 
> I'll also prepare a CL to solve the Rustix/Android issue..

Thanks! I've fixed the pending issues now.

gurchetansingh commented:
Apologies once again for the delay!  Rustix patch almost complete, undergoing testing internally.  In the meantime, can you send the `mesa3d_util` change to upstream Mesa?  

ETA for new rutabaga crates.io release at this pace: ~3 weeks

Thank you!

dorindabassey commented:
> Apologies once again for the delay! Rustix patch almost complete, undergoing testing internally. In the meantime, can you send the `mesa3d_util` change to upstream Mesa?
> 
> ETA for new rutabaga crates.io release at this pace: ~3 weeks
> 
> Thank you!

NO worries! Thank you! Submitted the PR here - https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/39653

gurchetansingh commented:
https://github.com/bytecodealliance/rustix/pull/1577 <-- Rustix pull request here, could take a few weeks to get a review tho

gurchetansingh commented:
@dorindabassey the Rustix MR is taking more time than expected.  

As a stopgap to make sure you're not blocked, I created a `libkrun` specific branch and associated crates.io releases:

https://github.com/magma-gpu/rutabaga_gfx/commits/libkrun
https://crates.io/crates/mesa3d_util/0.1.76-libkrun.0
https://crates.io/crates/rutabaga_gfx/0.1.76-libkrun.0

This is similar to the tricks used to support `chromeos` too.  Definitely will merge this into `main` once we have `rustix` issues fixed, but in the meantime feel free to use the `libkrun` crates.io releases to land the rest of the changes you need.

dorindabassey commented:
> @dorindabassey the Rustix MR is taking more time than expected.
> 
> As a stopgap to make sure you're not blocked, I created a `libkrun` specific branch and associated crates.io releases:
> 
> https://github.com/magma-gpu/rutabaga_gfx/commits/libkrun https://crates.io/crates/mesa3d_util/0.1.76-libkrun.0 https://crates.io/crates/rutabaga_gfx/0.1.76-libkrun.0
> 
> This is similar to the tricks used to support `chromeos` too. Definitely will merge this into `main` once we have `rustix` issues fixed, but in the meantime feel free to use the `libkrun` crates.io releases to land the rest of the changes you need.

Thanks! This would be helpful, also the Mesa MR is still pending.

gurchetansingh commented:
> also the Mesa MR is still pending

Ack, but merging that would also break Android unfortunately, and it's much less easier/useful to create branches for Mesa.

valpackett commented:
Sorry for dragging this along even more :) but one extra thing I'm currently working on (for libkrun) is D-Bus over virtgpu, so we could have D-Bus integration with FD-passing (for XDG portals that require it). Camera and screen capture pass PipeWire remotes, but printing and file transfer for drag&drop/copy&paste pass regular file fds. Well, printing does, but file transfer passes `O_PATH` descriptors (despite the document-portal preferring "normal" ones now). And that's what I wanted to discuss here:

this current `type VirtioFsTable = Arc<Mutex<Map<VirtioFsKey, File>>>` won't work for `O_PATH` fds, as in that case there's no handle at all, just the inode number. So I'd have to add a *different* map/thingy to be able to look up inodes.

But anyway *also* this whole `Mutex<Map<…>>` is straight out of libkrun's implementation details and a different VMM might use a different data structure to store the FDs. (And maybe message-passing instead of mutex-based sharing for concurrency!)

Can we instead make a trait like

```rust
pub trait VirtioFsLookup {
    async fn get_open_file(fs_id: u64, handle: u64) -> Result<File>;
}
```

and then store an `Option<dyn VirtioFsLookup + Send + Sync>` or something…?

(And then conveniently I'll be able to extend it with a method that returns an `O_PATH` "File" based on an inode instead of a handle)

dorindabassey commented:
> Sorry for dragging this along even more :) but one extra thing I'm currently working on (for libkrun) is D-Bus over virtgpu, so we could have D-Bus integration with FD-passing (for XDG portals that require it). Camera and screen capture pass PipeWire remotes, but printing and file transfer for drag&drop/copy&paste pass regular file fds. Well, printing does, but file transfer passes `O_PATH` descriptors (despite the document-portal preferring "normal" ones now). And that's what I wanted to discuss here:
> 
> this current `type VirtioFsTable = Arc<Mutex<Map<VirtioFsKey, File>>>` won't work for `O_PATH` fds, as in that case there's no handle at all, just the inode number. So I'd have to add a _different_ map/thingy to be able to look up inodes.
> 
> But anyway _also_ this whole `Mutex<Map<…>>` is straight out of libkrun's implementation details and a different VMM might use a different data structure to store the FDs. (And maybe message-passing instead of mutex-based sharing for concurrency!)
> 
> Can we instead make a trait like
> 
> ```rust
> pub trait VirtioFsLookup {
>     async fn get_open_file(fs_id: u64, handle: u64) -> Result<File>;
> }
> ```
> 
> and then store an `Option<dyn VirtioFsLookup + Send + Sync>` or something…?
> 
> (And then conveniently I'll be able to extend it with a method that returns an `O_PATH` "File" based on an inode instead of a handle)

I'm currently working on implementing the trait-based approach you suggested, though I made some adjustments from your proposal:
 pub trait VirtioFsLookup: Send + Sync {
     fn get_open_file(&self, fs_id: u64, handle: u64) -> RutabagaResult<File>;
 }
I decided to make it sync because The codebase has no async runtime, and the call site in AtomicMemorySentinelManager::create_watcher is synchronous. Is this ok for you? or do you prefer the async version?
For the O_PATH support, we can extend the trait with an optional method for inode lookups.
but I wonder if this is something to do in this PR, or create a new PR for it?

valpackett commented:
@dorindabassey sync is fine for everything right now, the possibility of using async is something I suggested for the very abstract possibility of other/future VMMs wanting to use async :)

For the inode method, just the possibility that I could add it to the trait in a future PR would be excellent.

mtjhrc commented:
Just a small suggestion, since this is switching to a more general trait anyway:

I think the `get_open_file` should return `OwnedDescriptor` instead of `File`.  That seems more semantically correct and the code converts it to an `OwnedDescriptor` immediately anyway. And maybe call it something like `get_exported_fd`?

dorindabassey commented:
> Just a small suggestion, since this is switching to a more general trait anyway:
> 
> I think the `get_open_file` should return `OwnedDescriptor` instead of `File`. That seems more semantically correct and the code converts it to an `OwnedDescriptor` immediately anyway. And maybe call it something like `get_exported_fd`?

Right, Thanks! the PR is opened on #44

gurchetansingh commented:
I cherry-picked all the libkrun changes into main a new MR: https://github.com/magma-gpu/rutabaga_gfx/pull/55, since maintaining two branches will only get more difficult with time.   The Rustix issue is solved via downstream patches.  PTAL and approve if you like it, thanks!
